File: Code/Chapter 8/C#/Exercise 8.7/Form1.cs
#region Using directives

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Windows.Forms;

#endregion

namespace Exercise_8_7
{
   partial class Form1 : Form
   {
      private static string dataFilePath = @"..\..\..\..\..\Data.Xml";
      private DataSet customerProducts;

      public Form1()
      {
         InitializeComponent();
      }

      private void Form1_Load(object sender, EventArgs e)
      {
         customerProducts = CreateDataSet.DataSetFiller.FillStrongDataSet(dataFilePath);
         foreach (DataRow dr in customerProducts.Tables["Customers"].Rows)
         {
            lbCustomers.Items.Add(dr["CustomerID"] + ":" + dr["FirstName"] + " " + dr["LastName"]);
         }
      }

      private void btnFilter2_Click(object sender, EventArgs e)
      {
         DataRow custProdRow;
         DataRow ProdRow;
         DataTable custProdTable = customerProducts.Tables["CustomerProducts"] ;
         lbProducts.Items.Clear();
         foreach (object item in lbCustomerProducts.Items)
         {
            int custProdId = (int)item;
            custProdRow = custProdTable.Rows.Find(custProdId);
            ProdRow = custProdRow.GetParentRow(customerProducts.Relations[0]);
            lbProducts.Items.Add(ProdRow["ProductName"]);
         }
      }

      private void btnFilter1_Click(object sender, EventArgs e)
      {
         if (lbCustomers.SelectedIndex == -1)
            return;
         DataRow selectedRow = customerProducts.Tables["Customers"].Rows[lbCustomers.SelectedIndex];
         DataRow[] childRows = selectedRow.GetChildRows(customerProducts.Relations[1]);
         lbCustomerProducts.Items.Clear();
         foreach (DataRow dr in childRows)
         {
            lbCustomerProducts.Items.Add(dr["CustomerProductID"]);
         }
      }
   }
}

File: Code/Chapter 8/VB.NET/Exercise 8.7/Form1.vb
Imports CreateDataSet
Imports System.Data

Public Class Form1
   Private dataFilePath As String = "..\..\..\..\Data.Xml"
   Private customerProducts As DataSet

   Private Sub btnFilter1_Click( _
     ByVal sender As System.Object, ByVal e As System.EventArgs) _
     Handles btnFilter1.Click
      If lbCustomers.SelectedIndex = -1 Then
         Return
      End If
      Dim selectedRow As DataRow = _
         customerProducts.Tables("Customers").Rows(lbCustomers.SelectedIndex)
      Dim childRows() As DataRow = _
         selectedRow.GetChildRows(customerProducts.Relations(1))
      lbCustomerProducts.Items.Clear()
      Dim dr As DataRow
      For Each dr In childRows
         lbCustomerProducts.Items.Add(dr("CustomerProductID"))
      Next
   End Sub
   Private Sub btnFilter2_Click( _
     ByVal sender As System.Object, ByVal e As System.EventArgs) _
     Handles btnFilter2.Click
      Dim custProdRow As DataRow
      Dim ProdRow As DataRow
      Dim custProdTable As DataTable = customerProducts.Tables("CustomerProducts")
      lbProducts.Items.Clear()
      Dim item As Object
      For Each item In lbCustomerProducts.Items
         Dim custProdId As Integer = CType(item, Integer)
         custProdRow = custProdTable.Rows.Find(custProdId)
         ProdRow = custProdRow.GetParentRow(customerProducts.Relations(0))
         lbProducts.Items.Add(ProdRow("ProductName"))
      Next
   End Sub

   Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
      customerProducts = CreateDataSet.DataSetFiller.FillStrongDataSet(dataFilePath)
      For Each dr As DataRow In customerProducts.Tables("Customers").Rows
         lbCustomers.Items.Add(dr("CustomerID") & ":" & dr("FirstName") & " " + dr("LastName"))
      Next
   End Sub
End Class
